pkg/flow/internal/controller/component_node_manager.go

```diff
@@ -0,0 +1,258 @@
+func (m *ComponentNodeManager) isCustomComponent(componentName string) bool {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+	namespace := strings.Split(componentName, ".")[0]
+	_, declareExists := m.declareNodes[namespace]
+	_, importExists := m.importNodes[namespace]
+	_, additionalDeclareContentExists := m.additionalDeclareContents[componentName]
+
+	return declareExists || importExists || additionalDeclareContentExists
+}
+
+// FindLocalDeclareNode returns the declareNode matching the declareLabel of the provided CustomComponentNode if present.
+func (m *ComponentNodeManager) FindLocalDeclareNode(cc *CustomComponentNode) (*DeclareNode, bool) {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+	declareNode, exist := m.declareNodes[cc.declareLabel]
+	return declareNode, exist
+}
+
+// FindImportedConfigNode returns the importNode matching the importLabel of the provided CustomComponentNode if present.
+func (m *ComponentNodeManager) FindImportedConfigNode(cc *CustomComponentNode) (*ImportConfigNode, bool) {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+	importNode, exist := m.importNodes[cc.importLabel]
+	return importNode, exist
+}
+
+// CustomComponentConfig represents the config needed by a custom component to load.
+type CustomComponentConfig struct {
+	declareContent            string            // represents the corresponding declare as plain string
+	additionalDeclareContents map[string]string // represents the additional declares that might be needed by the component to build custom components
+}
+
+// NewCustomComponentConfig creates a new CustomComponentConfig.
+func NewCustomComponentConfig(declareContent string, additionalDeclareContents map[string]string) *CustomComponentConfig {
+	return &CustomComponentConfig{
+		declareContent:            declareContent,
+		additionalDeclareContents: additionalDeclareContents,
+	}
+}
+
+// GetCustomComponentConfig returns the custom component config for a given custom component or an error if not found.
+func (m *ComponentNodeManager) GetCustomComponentConfig(cc *CustomComponentNode) (*CustomComponentConfig, error) {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+	var customComponentConfig *CustomComponentConfig
+	var err error
+	if cc.importLabel == "" {
+		customComponentConfig = m.getCustomComponentConfigFromLocalDeclares(cc)
+		if customComponentConfig == nil {
+			customComponentConfig = m.getCustomComponentConfigFromParent(cc)
+		}
+	} else {
+		customComponentConfig, err = m.getCustomComponentConfigFromImportedDeclares(cc)
+		if err != nil {
+			return customComponentConfig, err
+		}
+		if customComponentConfig == nil {
+			customComponentConfig = m.getCustomComponentConfigFromParent(cc)
+			// Custom components that receive their config from imported declares in a parent controller can only access the imported declares coming from the same import.
+			customComponentConfig.additionalDeclareContents = filterAdditionalDeclareContents(cc.importLabel, customComponentConfig.additionalDeclareContents)
+		}
```

Comment:
I'm finding this part of the code hard to keep all of the moving bits in my head, mainly around the three different ways you can get the definition for a custom component and how the "declare contents" is being passed around. 

While reading this, it feels like this could use a tree structure to represent the scope: 

* A scope for components declared in the local module (i.e., not imported).
  * This scope will have parents if you are inside a deeply nested declare component.   
* A scope for components per import statement. 

If we found a way to represent the declares as a tree, it might simplify the logic here into becoming: 

1. Find out what scope the component is in; local scope or imported scope
2. Search that scope for the component. 

I don't have the full answer here for how to design this, mainly because I'm still having a hard time following how the declare contents get passed around, but I _think_ a tree structure would eliminate that need. 

This also feeds into my comment about trying to reuse the ComponentRegistry interface; there might be a way to replace that with the tree scope idea and to have one general way of getting a component from a name given a current scope.

Reply:
I understand that this part of the code can be a bit daunting. I tried to have a descriptive function name but the verbosity might bring more confusion.

The logic is basically: A custom component needs its definition and additional definitions for nested custom components.
- If it's from a local declare, it checks the declare nodes, if not found then it checks the extra content passed by the parent.
- If it's from an imported declare, it checks the import nodes, if not found then it checks the extra content passed by the parent. 

Once it has its template, it must retrieve the additional declares needed for its nested custom components:
- If it's from a local declare, it can get the additional content from other local declares / imports / parent
- If it's from an imported declare, it can only get additional content from the corresponding import

The only trick is the prefix part for the import node to restrict the scope.

From the past discussions that we had on the topic I believe that we wanted to keep modules isolated (not leaking through parents or siblings). This approach follows this principle: you pass the config + any additional config needed and the module has everything it needs to run. I tried to follow what we wrote down during the last brainstorming session:
```
Loader changes
Scope lookup table for what namespaces are imported 
- Constructed on ApplyConfig 
- Has a hierarchy; nested loaders may have parent lookup table. 
```

That being said, I'm not against trying a new approach, especially if this one is not convincing enough or if it feels that there should be something better. The tree structure sounds like a good idea. Here is how I would see it:

```
type Scope struct {
  parent Scope
  declares map[string]string // customComponentName: template
  imports map[string]Scope // importNamespace: importScope
}
```
Instead of passing the additionalDeclaresContent down we can pass this scope object.
- We don't really need Declare Nodes, the content of the declare block can be directly set to the Scope
- onContentChange, ImportConfigNodes will need to update the Scope and the import hierarchy would need to be correctly mapped (if import A has an import B, then root Scope is Scope{imports["A"] = Scope{imports["B"]}}

For every custom component:
-if it's an instance of a local declare, you search the declares from the scope and continue up the parents till you find it. Once you have it, you can pass the current scope to the custom component so that it creates a new scope and reference the passed scope as a parent
- if it's an instance of an imported declare, you search the imports from the scope and continue up the parents till you find it. Once you have it, you can pass the scope corresponding to the import to the custom component so that it creates a new scope and reference the passed scope as a parent.

With this approach, the custom components are not isolated anymore. We may have situations where a child tries to access content from the parent scope while the parent is updating its scope. That will require extra care.

What do you think about this second approach?
Unless I'm missing something I should be able to implement it fairly quickly. I think that we should go for it

Reply:
> I see, my bad I misunderstood that you meant a tree in this case, I thought the hierarchy was that just parents passing down the map to children, not passing a tree of lookup tables.


No worries, I didn't mean anything bad by it, just pointing out that we're aligned with what you proposed :)